database/buildSchema/45_activity_log.sql

```diff
@@ -0,0 +1,383 @@
+CREATE TYPE public.event_type AS ENUM (
+    'STAGE',
+    'STATUS',
+    'OUTCOME',
+    'ASSIGNMENT',
+    'REVIEW',
+    'REVIEW_DECISION',
+    'PERMISSION' -- This type not (necessarily) tied to an application
+);
+
+CREATE TABLE public.activity_log (
+    id serial PRIMARY KEY,
+    type public.event_type NOT NULL,
+    value varchar NOT NULL,
+    timestamp timestamptz DEFAULT CURRENT_TIMESTAMP NOT NULL,
+    application_id integer REFERENCES public.application (id) ON DELETE CASCADE,
+    "table" varchar NOT NULL,
+    record_id integer,
+    details jsonb NOT NULL DEFAULT '{}'
+);
+
+-- Make an index on the application_id field, since this is the one it will be
+-- searched by most often
+CREATE INDEX activity_log_application_index ON activity_log (application_id);
+
+-- TRIGGERS and FUNCTIONS for updating activity log
+-- STAGE changes
+CREATE OR REPLACE FUNCTION public.stage_activity_log ()
+    RETURNS TRIGGER
+    AS $application_event$
+DECLARE
+    stage_num integer;
+    stage_name varchar;
+    stage_col varchar;
+BEGIN
+    SELECT
+        number,
+        title,
+        colour INTO stage_num,
+        stage_name,
+        stage_col
+    FROM
+        template_stage
+    WHERE
+        id = NEW.stage_id;
+    INSERT INTO public.activity_log (type, value, application_id, "table", record_id, details)
+        VALUES ('STAGE', stage_name, NEW.application_id, TG_TABLE_NAME, NEW.id, json_build_object('stage', json_build_object('number', stage_num, 'name', stage_name, 'colour', stage_col)));
+    RETURN NULL;
+END;
+$application_event$
+LANGUAGE plpgsql;
+
+CREATE TRIGGER stage_activity_trigger
+    AFTER INSERT ON public.application_stage_history
+    FOR EACH ROW
+    EXECUTE FUNCTION stage_activity_log ();
+
+-- STATUS changes
+CREATE OR REPLACE FUNCTION public.status_activity_log ()
+    RETURNS TRIGGER
+    AS $application_event$
+DECLARE
+    app_id integer;
+    prev_status varchar;
+BEGIN
+    app_id = (
+        SELECT
+            application_id
+        FROM
+            application_stage_history
+        WHERE
+            id = NEW.application_stage_history_id);
+    prev_status = (
+        SELECT
+            status
+        FROM
+            application_status_history
+        WHERE
+            application_id = app_id
+            AND is_current = TRUE);
+    INSERT INTO public.activity_log (type, value, application_id, "table", record_id, details)
+        VALUES ('STATUS', (
+                CASE WHEN NEW.status = 'DRAFT'
+                    AND prev_status IS NULL THEN
+                    'Started'
+                WHEN NEW.status = 'SUBMITTED'
+                    AND prev_status = 'COMPLETED' THEN
+                    'New Stage'
+                WHEN NEW.status = 'DRAFT'
+                    AND prev_status = 'CHANGES_REQUIRED' THEN
+                    'Re-started'
+                    -- HOW TO RECOGNISE "RE-SUBMITTED" CASE ?
```

Comment:
No, because there'll be a "CHANGES_REQUESTED" status in between SUBMITTED and DRAFT. 

I've done some front-end logic to deduce this anyway, probably easier in there: https://github.com/openmsupply/application-manager-web-app/pull/1091

Reply:
Oh, from my understanding the review would be 'DRAFT' before submitting, but it would be difficult to differentiate different cases. It would be (previously to DRAFT) wither PENDING or CHANGE_REQUESTED, so probably better to keep as you have done so far. I'll have a look in it next.

Reply:
All this "interpretation" logic is now in one place -- the "eventInterpretation" file, which is part of the "useTimeline" hook. So it should be easier to check what's going on for each type of event and amend it if necessary.